src/raw-value.ts

```diff
@@ -239,6 +263,78 @@ export function reverseProxyFactory(env: MembraneBroker) {
             }
             return env.getRawValue(sec);
         }
+        has(shadowTarget: ReverseShadowTarget, key: PropertyKey): boolean {
+            const { target } = this;
+            // todo: protect against leaking symbols
+            return key in target;
+        }
+        ownKeys(shadowTarget: ReverseShadowTarget): PropertyKey[] {
+            const { target } = this;
+            // todo: protect against leaking symbols
+            return ownKeys(target);
+        }
+        isExtensible(shadowTarget: ReverseShadowTarget): boolean {
+            if (!ReflectIsExtensible(shadowTarget)) {
+                return false;
+            }
+            const { target } = this;
+            if (!ReflectIsExtensible(target)) {
+                lockShadowTarget(shadowTarget, target);
+                return false;
+            }
+            return true;
+        }
+        getOwnPropertyDescriptor(shadowTarget: ReverseShadowTarget, key: PropertyKey): PropertyDescriptor | undefined {
+            const { target } = this;
+            // todo: leaking key as symbol
+            let rawDesc = ReflectGetOwnPropertyDescriptor(shadowTarget, key);
+            if (!isUndefined(rawDesc)) {
+                return rawDesc;
+            }
+            const secDesc = ReflectGetOwnPropertyDescriptor(target, key);
+            if (isUndefined(secDesc)) {
+                return secDesc;
+            }
+            rawDesc = getRawDescriptor(secDesc);
+            if (secDesc.configurable == false) {
+                // updating the descriptor to non-configurable on the shadow
+                ReflectDefineProperty(shadowTarget, key, rawDesc);
+            }
+            return rawDesc;
+        }
+        getPrototypeOf(shadowTarget: ReverseShadowTarget): RawValue {
+            const { target } = this;
+            return env.getRawValue(ReflectGetPrototypeOf(target));
+        }
+        setPrototypeOf(shadowTarget: ReverseShadowTarget, prototype: RawValue): boolean {
+            const { target } = this;
+            return ReflectSetPrototypeOf(target, env.getSecureValue(prototype));
+        }
+        preventExtensions(shadowTarget: ReverseShadowTarget): boolean {
+            if (ReflectIsExtensible(shadowTarget)) {
+                const { target } = this;
+                if (!ReflectPreventExtensions(target)) {
+                    // if the target is a proxy manually created in the sandbox, it might reject
+                    // the preventExtension call, in which case we should not attempt to lock down
+                    // the shadow target.
+                    return false;
+                }
+                lockShadowTarget(shadowTarget, target);
```

Comment:
In the `isExtensible` trap you lock the `shadowTarget` if `target` prevents extensions. In the bit above you exit early before locking the `shadowTarget` with a comment about not wanting to lock it. Does the same reasoning extend to the `isExtensible` trap?

Reply:
hmm, I think we are good there, let me articulate:

```js
        isExtensible(shadowTarget: ReverseShadowTarget): boolean {
            if (!ReflectIsExtensible(shadowTarget)) {
                return false;
            }
            const { target } = this;
            if (!ReflectIsExtensible(target)) {
                lockShadowTarget(shadowTarget, target);
                return false;
            }
            return true;
        }
```

If `shadowTarget` is non-extensible, the only way that could happen is because it was already locked down, and the only way it can be locked down is because either this method is called before, or preventExtension trap.

If the `shadowTarget` is extensible, it only means that either the target is extensible, or we haven't check it yet, which we do in the next condition.

Does this make sense?

Reply:
Makes sense 😸